packages/loopring_v3/circuit/Gadgets/MathGadgets.h

```diff
@@ -362,49 +549,51 @@ class TernaryGadget : public GadgetT
     }
 };
 
-// (A <(=) B)
-class LeqGadget : public GadgetT
+// b ? A : B
+class ArrayTernaryGadget : public GadgetT
 {
 public:
-    VariableT _lt;
-    VariableT _leq;
-    libsnark::comparison_gadget<ethsnarks::FieldT> comparison;
 
-    LeqGadget(
+    std::vector<TernaryGadget> results;
+    VariableArrayT res;
+
+    ArrayTernaryGadget(
         ProtoboardT& pb,
-        const VariableT& A,
-        const VariableT& B,
-        const size_t n,
+        const VariableT& b,
+        const VariableArrayT& x,
+        const VariableArrayT& y,
         const std::string& prefix
     ) :
-        GadgetT(pb, prefix),
-
-        _lt(make_variable(pb, 1, FMT(prefix, ".lt"))),
-        _leq(make_variable(pb, 1, FMT(prefix, ".leq"))),
-        comparison(pb, n, A, B, _lt, _leq, FMT(prefix, ".A <(=) B"))
-    {
-        // The comparison gadget is only guaranteed to work correctly on values in the field capacity - 1
-        assert(n <= NUM_BITS_FIELD_CAPACITY - 1);
-    }
-
-    const VariableT& lt() const
+        GadgetT(pb, prefix)
     {
-        return _lt;
+        assert(x.size() == y.size());
+        results.reserve(x.size());
+        for (unsigned int i = 0; i < x.size(); i++)
+        {
+            results.emplace_back(TernaryGadget(pb, b, x[i], y[i], FMT(prefix, ".results")));
+            res.emplace_back(results.back().result());
+        }
     }
 
-    const VariableT& leq() const
+    void generate_r1cs_witness()
     {
-        return _leq;
+        for (unsigned int i = 0; i < results.size(); i++)
+        {
+            results[i].generate_r1cs_witness();
+        }
     }
 
-    void generate_r1cs_witness()
+    void generate_r1cs_constraints()
```

Comment:
maybe we also add a 'enforceBitness' parameter and check it as in TernaryGadget

Reply:
Sure, the assumption here is that `b` is boolean because otherwise things don't make sense at all, but can't hurt to add.